File: inference_framework/inference_framework/schema_validator.py
import json
from pathlib import Path
from typing import Dict, Any, Type, List, Optional
from pydantic import BaseModel, Field, create_model, UUID4
import logging

logger = logging.getLogger(__name__)

class SchemaValidator:
    """Handles dynamic Pydantic model creation from schema files and validation."""
    
    def __init__(self, schemas_path: str = "schemas"):
        self.schemas_path = Path(schemas_path)
        self.models: Dict[str, Type[BaseModel]] = {}
        self._load_schemas()
    
    def _load_schemas(self):
        """Load all schema definitions from JSON files."""
        if not self.schemas_path.exists():
            logger.warning(f"Schemas directory not found: {self.schemas_path}")
            return
        for schema_file in self.schemas_path.glob("*.json"):
            schema_name = schema_file.stem
            with open(schema_file, 'r') as f:
                schema_def = json.load(f)
            self.models[schema_name] = self._create_pydantic_model(schema_name, schema_def)
            logger.info(f"Loaded schema: {schema_name}")
    
    def _create_pydantic_model(self, name: str, schema_def: Dict[str, Any]) -> Type[BaseModel]:
        """Dynamically create Pydantic model from schema definition."""
        fields = {}
        field_types = {
            "str": str,
            "int": int,
            "float": float,
            "bool": bool,
            # Add more as needed
        }
        
        for field_name, field_type_str in schema_def.get("fields", {}).items():
            field_type = field_types.get(field_type_str, str)  # default to str
            if field_name == "event_id":
                field_type = UUID4  # special for UUID
            # Make optional unless required
            required = field_name in schema_def.get("required", [])
            if required:
                fields[field_name] = (field_type, Field(...))
            else:
                fields[field_name] = (Optional[field_type], Field(None))
        
        model = create_model(name, **fields)
        return model
    
    def get_model(self, schema_name: str) -> Optional[Type[BaseModel]]:
        """Get Pydantic model for schema."""
        return self.models.get(schema_name)
    
    def validate_message(self, message: Dict[str, Any], schema_name: str = "app_log_type_login") -> Optional[BaseModel]:
        """Validate message against schema. Return validated model or None if invalid."""
        model_cls = self.get_model(schema_name)
        if not model_cls:
            logger.error(f"Schema not found: {schema_name}")
            return None
        try:
            # Parse and validate
            validated = model_cls(**message)
            return validated
        except Exception as e:
            logger.warning(f"Validation failed for message: {str(e)}")
            return None

File: inference_framework-java/src/main/java/com/inferenceframework/schema/SchemaValidator.java
package com.inferenceframework.schema;  // schema module

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.networknt.schema.JsonSchema;
import com.networknt.schema.JsonSchemaFactory;
import com.networknt.schema.SpecVersion;
import com.networknt.schema.ValidationMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;

/**
 * Handles dynamic schema validation using JSON Schema Validator library (networknt).
 * Converts custom schema files to standard JSON Schema format for validation.
 * Appropriate library for JSON schema validation in Java.
 */
public class SchemaValidator {
    private static final Logger logger = LoggerFactory.getLogger(SchemaValidator.class);

    private final String schemasPath;
    private final ObjectMapper mapper;
    private final Map<String, JsonSchema> schemas;  // schema_name -> JsonSchema

    /**
     * Constructor loads schemas from directory.
     */
    public SchemaValidator(String schemasPath) {
        this.schemasPath = schemasPath;
        this.mapper = new ObjectMapper();
        this.schemas = new HashMap<>();
        loadSchemas();
    }

    /**
     * Default constructor.
     */
    public SchemaValidator() {
        this("schemas");
    }

    /**
     * Load all schema definitions from JSON files and convert to JsonSchema.
     */
    private void loadSchemas() {
        Path dir = Paths.get(schemasPath);
        if (!Files.exists(dir)) {
            logger.warn("Schemas directory not found: {}", schemasPath);
            return;
        }
        try {
            Files.list(dir)
                .filter(p -> p.toString().endsWith(".json"))
                .forEach(this::loadSchema);
        } catch (IOException e) {
            logger.error("Failed to list schemas", e);
        }
    }

    /**
     * Load and convert one schema file to standard JSON Schema for validation.
     */
    private void loadSchema(Path schemaFile) {
        String schemaName = schemaFile.getFileName().toString().replace(".json", "");
        try {
            // Read custom schema def
            JsonNode customSchemaDef = mapper.readTree(schemaFile.toFile());
            // Convert to standard JSON Schema (Draft 2020-12 via lib)
            JsonNode jsonSchemaNode = convertToJsonSchema(customSchemaDef);
            JsonSchemaFactory factory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V202012);
            JsonSchema schema = factory.getSchema(jsonSchemaNode);
            schemas.put(schemaName, schema);
            logger.info("Loaded schema: {}", schemaName);
        } catch (Exception e) {
            logger.error("Failed to load schema: {}", schemaName, e);
        }
    }

    /**
     * Convert custom schema format to standard JSON Schema object.
     * Custom: {"fields": {"field": "type"}, "required": [...]}
     * Standard: {"type": "object", "properties": {...}, "required": [...]}
     */
    private JsonNode convertToJsonSchema(JsonNode customDef) {
        ObjectMapper om = new ObjectMapper();
        Map<String, Object> stdSchema = new HashMap<>();
        stdSchema.put("type", "object");
        stdSchema.put("additionalProperties", true);  // allow extra

        Map<String, Object> properties = new HashMap<>();
        JsonNode fields = customDef.get("fields");
        if (fields != null && fields.isObject()) {
            fields.fields().forEachRemaining(entry -> {
                String typeStr = entry.getValue().asText();
                Map<String, Object> propType = new HashMap<>();
                propType.put("type", mapType(typeStr));
                properties.put(entry.getKey(), propType);
            });
        }
        stdSchema.put("properties", properties);

        // Required fields
        JsonNode required = customDef.get("required");
        if (required != null && required.isArray()) {
            List<String> reqList = new ArrayList<>();
            required.forEach(n -> reqList.add(n.asText()));
            stdSchema.put("required", reqList);
        }

        return om.valueToTree(stdSchema);
    }

    /**
     * Map simple type strings to JSON Schema types.
     */
    private String mapType(String typeStr) {
        switch (typeStr.toLowerCase()) {
            case "str":
            case "string": return "string";
            case "int":
            case "integer": return "integer";
            case "float":
            case "number": return "number";
            case "bool":
            case "boolean": return "boolean";
            default: return "string";
        }
    }

    /**
     * Get schema for name.
     */
    public JsonSchema getSchema(String schemaName) {
        return schemas.get(schemaName);
    }

    /**
     * Validate message (Map/JsonNode) against schema.
     * Return validated map (as-is if valid) or null if invalid.
     * For compatibility with Python version.
     */
    public Map<String, Object> validateMessage(Map<String, Object> message, String schemaName) {
        JsonSchema schema = getSchema(schemaName);
        if (schema == null) {
            logger.error("Schema not found: {}", schemaName);
            return null;
        }
        try {
            // Convert message to JsonNode for validation
            JsonNode jsonNode = mapper.valueToTree(message);
            Set<ValidationMessage> errors = schema.validate(jsonNode);
            if (!errors.isEmpty()) {
                logger.warn("Validation failed for message: {}", errors);
                return null;
            }
            // Valid: return original map
            return message;
        } catch (Exception e) {
            logger.warn("Validation error: {}", e.getMessage());
            return null;
        }
    }
}